### PR TITLE
Redesign of the drugorder widget

### PR DESCRIPTION
**Pull request by manuelroemer**
Includes the changes which lead to the current redesign state.
There are still some open TODOs which will (hopefully) be resolved soon, but the current version is nontheless functional for the most part.
Feel free to rip it apart. ;)

Screens:
grafik
grafik
grafik

### CODE REVIEW

**Comment by brandones** [comment]
Can we get some screenshots or a gif?

File: jest.config.js [hunk 1]
```diff
@@ -4,6 +4,7 @@ module.exports = {
   },
   moduleNameMapper: {
     '\\.(css)$': 'identity-obj-proxy',
+    '\\.(scss)$': 'identity-obj-proxy',
```
Comment (FlorianRappl) [inline jest.config.js]: Could also be one rule `\\.(s?css)$`

File: src/api/api.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { openmrsFetch } from '@openmrs/esm-api';
+import { careSetting } from '../constants';
+import { Order, OrderPost } from '../types/order';
+
+const DURATION_UNITS_CONCEPT = '1732AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA';
```
Comment (FlorianRappl) [inline src/api/api.ts]: Personally I prefer `durationUnitsConcept` here. The all caps etc. style is something that was used exclusively in C to not mix macros with other parts.

Comment (FlorianRappl) [reply src/api/api.ts]: (one use of a macro was a compile-time constant, i.e., just a constant. that concept does not exist in JS - the `const` is not a compile-time constant, but rather a variable that cannot be reassigned)

File: src/components/medications-details-table.component.tsx [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+  TableHeader,
+  TableRow,
+  TableToolbar,
+  TableToolbarContent,
+} from 'carbon-components-react';
+import { getDosage } from '../utils/get-dosage';
+import dayjs from 'dayjs';
+import { useTranslation } from 'react-i18next';
+import { Add16 } from '@carbon/icons-react';
+import { capitalize } from 'lodash-es';
```
Comment (FlorianRappl) [inline src/components/medications-details-table.component.tsx]: Can we check if this gets bundled efficiently? Potentially, `lodash-es/capitalize` (or similar) would be more efficient.

File: src/components/medications-details-table.component.tsx [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                        isSortable: header.isSortable,
+                      })}>
+                      {header.header}
+                    </TableHeader>
+                  ))}
+                  <TableHeader />
+                </TableRow>
+              </TableHead>
+              <TableBody>
+                {rows.map((row, rowIndex) => (
```
Comment (FlorianRappl) [inline src/components/medications-details-table.component.tsx:134]: This here leaves room for performance improvements.

We can discuss this tomorrow. Not necessary for this PR though.

File: src/order-basket/drug-ordering.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+const careSetting = '6f0c9a92-6f24-11e3-af88-005056821db0';
+const orderer = 'e89cae4a-3cb3-40a2-b964-8b20dda2c985';
+
+export async function orderDrugs(
+  orderBasketItems: Array<OrderBasketItem>,
+  patientUuid: string,
+  abortController: AbortController,
+) {
+  const dtos = medicationOrderToApiDto(orderBasketItems, patientUuid);
+  const erroredItems = [];
```
Comment (FlorianRappl) [inline src/order-basket/drug-ordering.ts]: This is `Array<any>`. Can we be more specific?

File: src/order-basket/drug-search.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import { getDrugByName } from '../api/api';
+import { getCommonMedicationByUuid } from '../api/common-medication';
+import { OrderBasketItem } from '../types/order-basket-item';
+import { daysDurationUnit } from '../constants';
+import _ from 'lodash-es';
```
Comment (FlorianRappl) [inline src/order-basket/drug-search.ts]: This imports the whole lodash-es package. Do we need this? Or just something specific?

File: src/order-basket/medication-order-form.component.tsx [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+  onSign,
+  onCancel,
+}: MedicationOrderFormProps) {
+  const { t } = useTranslation();
+  const [orderBasketItem, setOrderBasketItem] = useState(initialOrderBasketItem);
+  const commonMedication = getCommonMedicationByUuid(orderBasketItem.drug.uuid);
+
+  return (
+    <>
+      <Header aria-label={''} className={styles.medicationDetailsHeader}>
```
Comment (FlorianRappl) [inline src/order-basket/medication-order-form.component.tsx]: `{''}` -> `""`

File: src/order-basket/medication-order-form.component.tsx [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+            <Column md={2}>
+              <NumberInput
+                id="durationInput"
+                label={t('duration', 'Duration')}
+                min={0}
+                value={orderBasketItem.duration}
+                helperText={t('noDurationHint', '0 indicates an indefinite duration.')}
+                onChange={e => {
+                  // @ts-ignore
+                  const newValue = Number(e.imaginaryTarget.value);
```
Comment (FlorianRappl) [inline src/order-basket/medication-order-form.component.tsx]: Or just `+e.imaginaryTarget.value`.

File: src/order-basket/medication-order-form.component.tsx [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+                <NumberInput
+                  id="quantityDispensed"
+                  helperText={t('pillsDispensed', 'Pills dispensed')}
+                  value={orderBasketItem.pillsDispensed}
+                  min={0}
+                  onChange={e => {
+                    setOrderBasketItem({
+                      ...orderBasketItem,
+                      // @ts-ignore
+                      pillsDispensed: Number(e.imaginaryTarget.value),
```
Comment (FlorianRappl) [inline src/order-basket/medication-order-form.component.tsx]: Same as above. BTW `+foo` is accepted in TS as a converter.

File: src/types/order.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
+import { OpenmrsResource } from './openmrs-resource';
+
+export type Order = {
```
Comment (FlorianRappl) [inline src/types/order.ts]: I'd prefer using `interface` for such constructs.

File: src/utils/get-dosage.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  let strengthUnits = strength.substring(i);
+  let dosage;
+
+  if (concentrationStartIndex >= 0) {
+    strengthUnits = strength.substring(i, concentrationStartIndex);
+    const j = strength.substring(concentrationStartIndex + 1).search(/\D/);
+    const concentrationQuantity = strength.substr(concentrationStartIndex + 1, j);
+    const concentrationUnits = strength.substring(concentrationStartIndex + 1 + j);
+    dosage =
+      doseNumber +
```
Comment (FlorianRappl) [inline src/utils/get-dosage.ts]: Hehe quite some concats here.

What you can do is to abuse an array.

```
dosage = [doseNumber, strengthUnits, `(${(doseNumber / strengthQuantity) * concentrationQuantity}`, `${concentrationUnits})`].join(' ');
```

File: src/utils/pagination.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import _ from 'lodash-es';
```
Comment (FlorianRappl) [inline src/utils/pagination.ts]:
```ts
import chunk from 'lodash-es/chunk';
```

**Review by FlorianRappl — COMMENTED** [review]
Great - wonderful job :beers:!

**Review by FlorianRappl — APPROVED** [review]
Great!

**Review by FlorianRappl — APPROVED** [review]

**Pull request merged**